elfo-dumper/src/config/dump_path.rs

```diff
@@ -0,0 +1,259 @@
+        match var {
+            Variable::Class => dest.push_str(vars.class),
+            Variable::Time => {
+                dest.push_str(&vars.ts.to_string());
+            }
+        }
+    }
+
+    pub(crate) fn render(&self, variables: TemplateVariables<'_>) -> String {
+        let mut output = String::with_capacity(self.template.len());
+        let mut offset = 0;
+
+        for Component { size, data } in self.components.iter().copied() {
+            let size = size as usize;
+            match data {
+                ComponentData::Variable(var) => {
+                    Self::expand_variable(var, &variables, &mut output);
+                }
+                ComponentData::Path => {
+                    output.push_str(&self.template[offset..offset + size]);
+                }
+            }
+
+            offset += size;
+        }
+
+        output
+    }
+}
+
+impl DumpPath {
+    fn parse_variable(var: &str) -> Option<Variable> {
+        use Variable as V;
+
+        Some(match var {
+            "time" => V::Time,
+            "class" => V::Class,
+            _ => return None,
+        })
+    }
+
+    fn push_path(of_size: &mut usize, to: &mut Vec<Component>) {
+        if *of_size != 0 {
+            to.push(Component {
+                size: *of_size as u16,
+                data: ComponentData::Path,
+            });
+            *of_size = 0;
+        }
+    }
+
+    /// Parse template.
+    fn parse(s: impl Into<String>) -> DumpPath {
+        let template = s.into();
+        let mut components = vec![];
+
+        let mut plain_size = 0;
+        let mut chunk = template.as_str();
+
+        loop {
+            // original = "xxx {yyy} zzz"
+            // before = "xxx "
+            // after = "yyy} zzz"
+            let Some((before, after)) = chunk.split_once('{') else {
+                // no more template variables.
+                plain_size += chunk.len();
+                chunk = "";
+                break;
+            };
+            plain_size += before.len();
+
+            // raw_variable = "yyy"
+            // after = " zzz"
+            let Some((raw_variable, after)) = after.split_once('}') else {
+                // 1 = {, which is not in "before".
+                plain_size += 1;
+                chunk = after;
+                continue;
+            };
+            let Some(variable) = Self::parse_variable(raw_variable) else {
+                // "{<variable>}"
+                plain_size += 2 + raw_variable.len();
+                chunk = after;
+                continue;
+            };
+
+            Self::push_path(&mut plain_size, &mut components);
+            components.push(Component {
+                // "{<variable>}"
+                size: raw_variable.len() as u16 + 2,
+                data: ComponentData::Variable(variable),
+            });
+            chunk = after;
+        }
+        if !chunk.is_empty() {
```

Comment:
This is not needed, since the loop can only exit through `chunk = ""`

Reply:
Interestingly, if I remove that branch, rust itself says that assignment in `chunk = ""` is not used (no reads after this operation). Kinda surprised how compiler sometimes not that dumb.